### PR TITLE
Implement vectorized base image augmentation layer w/ `Grayscale()` example

### PR DESCRIPTION
**Pull request by LukeWood**
We have a ton of performance issues with KPLs due to auto vectorization not working with tf.image ops.

Basically my proposed approach is as follows:

- replace KPLs that can be vectorized fully with  VectorizedBaseImageAugmentation subclasses

Fixes: https://github.com/keras-team/keras-cv/issues/581

---

Caveats of this approach:
- a bit less flexibility (i.e. in the case image augmentation could be vectorized, but bounding box augmentation could be not)
- loss of ragged support
- perhaps a bit convoluted implementation wise to have two base layers

### CODE REVIEW

**Comment by bhack** [comment]
What about the product decision at https://github.com/tensorflow/tensorflow/issues/55639#issuecomment-1310150046?

**Comment by LukeWood** [comment]
> What about the product decision at [tensorflow/tensorflow#55639 (comment)](https://github.com/tensorflow/tensorflow/issues/55639#issuecomment-1310150046)?

Curious @fchollet 's thoughts on this.  Its impossible to vectorize this one in TensorFlow and keep transformations independent from one another.  Do you think a slowdown is acceptable if this is the case?

**Comment by bhack** [comment]
The bottleneck is with any API that doesn't have a batch dimension on param that you want to randomize or change x element.

So it is like if you need to loop over the batch calling the same op with a different parameter.  

I don't know if this could happen or not in the near future with the new XLA/MLIR birdge design and its genenerated kernels but right now not with vectorized map and I think not with the current bridge.

**Comment by bhack** [comment]
https://github.com/tensorflow/tensorflow/pull/55335#discussion_r832778484

@jpienaar Can ops like this one be fused in a loop without breaking the raw ops API?

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:1]: nit: 2023

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+BOUNDING_BOXES = "bounding_boxes"
+KEYPOINTS = "keypoints"
+SEGMENTATION_MASKS = "segmentation_masks"
+
+IS_DICT = "is_dict"
+BATCHED = "batched"
+USE_TARGETS = "use_targets"
+
+
+class VectorizedBaseImageAugmentationLayer(
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:37]: Is there a reason this can't subclass BaseImageAugmentationLayer?

File: keras_cv/layers/preprocessing/random_contrast.py [hunk 3]
```diff
@@ -56,33 +55,44 @@ class RandomContrast(BaseImageAugmentationLayer):
...
-        return self.factor()
-
-    def augment_image(self, image, transformation, **kwargs):
-        contrast_factor = transformation
-        output = tf.image.adjust_contrast(image, contrast_factor=contrast_factor)
+    def get_random_transformation_batch(self, batch_size, **kwargs):
+        return
+
+    def augment_images(self, images, transformations, **kwargs):
+        # Note: we skip get_random_transformation_batch() in RandomContrast because the
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/random_contrast.py:77]: This comment seems like it belongs in the `get_random_transformation_batch` impl

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:37]: discussed a bit offline but to summarize for community members:

they barely share functionality in reality despite looking like they may.

**Comment by LukeWood** [comment]
@jbischof @ianstenbit this is ready for a real review now.  We can incrementally migrate layers to vectorized layers using this approach.

**Comment by bhack** [comment]
Before we merge I hope that we could clarify the point about the product decision at:

https://github.com/keras-team/keras-cv/issues/1256#issuecomment-1376134450

As It was since the first keras-cv commits that I was talking about the performance overhead of this design with `tf.image` API.

**Comment by LukeWood** [comment]
> Before we merge I hope that we could clarify the point about the product decision at:
> 
> [#1256 (comment)](https://github.com/keras-team/keras-cv/issues/1256#issuecomment-1376134450)
> 
> As It was since the first keras-cv commits that I was talking about the performance overhead of this design with `tf.image` API.

What are we hoping to have clarified?  This change solves the performance issues of grayscale with no regression of functionality.  Item-wise augmentation is still deemed important.

**Comment by bhack** [comment]
> What are we hoping to have clarified? This change solves the performance issues of grayscale with no regression of 

It is not only about grayscale as this related to many ops. An August not update list is at https://github.com/keras-team/keras-cv/issues/291#issuecomment-1230707124

**Comment by LukeWood** [comment]
> > What are we hoping to have clarified? This change solves the performance issues of grayscale with no regression of
> 
> It is not only about grayscale as this is related to many ops. An August not update list is at [#291 (comment)](https://github.com/keras-team/keras-cv/issues/291#issuecomment-1230707124)

My goal in this PR is to provide an abstraction to allow us to optimize the performance of layers that are not being converted by pfor: https://github.com/keras-team/keras-cv/issues/291#issuecomment-1230707124

Does that answer the question?

**Comment by bhack** [comment]
> Does that answer the question?

This was the same topic one year ago but we didn't care about the topic also if we was in a design phase:
https://github.com/keras-team/keras-cv/issues/455#issuecomment-1137111935

So what have changed in 1 year that instead we could not care in the design phase with all the explicit feedback and benchmarks that I've sent?

> not being converted by pfor

As you can read in many place and also in that comment it is not the only point for the failure/fallback pfor conversion as we have also missing converters.

As an extra note:
after all these months it seems that we don't have a tf.image API owners and probably also a pfor API owner as all the related tickets are without any other comet or action point.

**Comment by LukeWood** [comment]
> > Does that answer the question?
> 
> This was the same topic one year ago but we didn't care about the topic also if we was in a design phase: [#455 (comment)](https://github.com/keras-team/keras-cv/issues/455#issuecomment-1137111935)
> 
> So what have changed in 1 year that instead we could not care in the design phase with all the explicit feedback and benchmarks that I've sent?
> 
> > not being converted by pfor
> 
> As you can read in many place and also in that comment it is not the only point for the failure/fallback of pfor conversion as we have also missing converters.
> 
> As an extra note: after all these months it seems that we don't have tf.image API owners and probably also pfor API owner as all the related tickets are without any other comment or action.

we do care about performance, and as you say there is no tf.image API owner or pfor API owner, so instead of for these issues to be fixed we'd like to offer vectorized implementations of these layers in the meantime.  Does that answer your question?

And to be clear, it was never that we didn't care about performance, it was simply that getting RandAugment working, OD augmentations working, and other product goals took priority.  Now that these are working I'm attempting to revisit these layers and optimize without major regressions in functionality.

**Comment by bhack** [comment]
>  it was never that we didn't care about performance

Honestly the impression for one year it was that we did not care about performance. I have not collected any comment that we recognized the problem also if submitted in multiple ticket and that we just need to postpone the design for resource constrain.

In any case better later then never, we will see the residual fallback after this migration.

About the PR:
Do we need to subclass `tf.keras.__internal__.layers.BaseRandomLayer` for every new policy?

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks Luke!

Lots of comments because this is a foundational piece and I want to make sure we get it really right.

File: keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+            result[SEGMENTATION_MASKS] = segmentation_masks
+
+        # preserve any additional inputs unmodified by this layer.
+        for key in inputs.keys() - result.keys():
+            result[key] = inputs[key]
+        return result
+
+    def call(self, inputs, training=True):
+        inputs = self._ensure_inputs_are_compute_dtype(inputs)
+        if training:
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py:228]: Good idea, done

File: keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+            else:
+                raise ValueError(
+                    "Image augmentation layers are expecting inputs to be "
+                    "rank 3 (HWC) or 4D (NHWC) tensors. Got shape: "
+                    f"{images.shape}"
+                )
+        else:
+            return inputs
+
+    def _format_inputs(self, inputs):
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py:242]: there are so many other layers on the base layer, and this will eventually replace the base layer if we finish migrating every layer.  

I don't think the case here is strong enough for this to be a subclass.

File: keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        metadata = {IS_DICT: True, USE_TARGETS: False}
+        if tf.is_tensor(inputs):
+            # single image input tensor
+            metadata[IS_DICT] = False
+            inputs = {IMAGES: inputs}
+
+        metadata[BATCHED] = inputs["images"].shape.rank == 4
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py:249]: actually no, you can call any of these layers with a single input.  The "batched" quality of this layer refers to an implementation detail, not that it must be called on batches.

File: keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        if inputs["images"].shape.rank == 3:
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py:250]: @ianstenbit see here, we expand dims if its operation in unbatched mode.

File: keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/batched_base_image_augmentation_layer.py:242]: actually this has some different behavior in it too - in the expand_dims() section below.

**Comment by LukeWood** [comment]
@ianstenbit all comments addressed.

Closed in favor of https://github.com/keras-team/keras-cv/pull/1373/files

**Comment by LukeWood** [comment]
> > it was never that we didn't care about performance
> 
> Honestly the impression for one year was that we did not care about performance. I have not collected any comment that we recognized the problem also if it was submitted and benchmarked in multiple tickets and that we just need to postpone the design for resource constrain.
> 
> In any case better later then never, we will see the residual fallback after this migration.
> 
> About the PR: Do we need to subclass `tf.keras.__internal__.layers.BaseRandomLayer` for every new policy?

```
Honestly the impression for one year was that we did not care about performance. I have not collected any comment that we recognized the problem also if it was submitted and benchmarked in multiple tickets and that we just need to postpone the design for resource constrain.
```

Resource constraints are real and we were hoping pfor would either be:

- fixed
- sufficiently fast

Neither turned out so we are now prioritizing this.  Please keep in mind resource constraints are a real issue, and they don't mean issues aren't important; they simply mean we have other things that are more pressing at the time.

**Comment by bhack** [comment]
> Neither turned out so we are now prioritizing this. Please keep in mind resource constraints are a real issue, and they don't mean issues aren't important; they simply mean we have other things that are more pressing at the time.

Generally I agree but it is a little bit different when we was still in the design phase and we had realtime feedback about how it will go to fail. 
If not it will be hard to interact clearly on the same page (mainly on PR) if we give priority/attention to f2f meeting.

Pfor could be not fixed if the underline design API of the ops we want to use in this library is not batch ready.

I think our approach was really out of design for Pfor from the early days. 

Then for missing converters it could be still fixed.

We cared about something (within the batch policy) with some theoretical validity but without an empirical verification about the faster convergence vs x epoch computational overhead.

Also as we have seen in another recent PR pytorch vision is not working within the batch so many reference models that we also train/replicare and the related reference papers were not trained with the within the batch policy at all.

**Comment by LukeWood** [comment]
> We cared about something (within the batch policy) with some theoretical validity but without an empirical verification about the faster convergence vs x epoch computational overhead.

This already matters in contrastive learning and any case where samples are compared within a batch.  This is a case where a silent error could really waste a users time someday - and we should avoid this if possible.

In the current structure we will likely be able to have both performance and numerical correctness.  I think this is the best of all worlds

**Comment by bhack** [comment]
> This already matters in contrastive learning and any case where samples are compared within a batch.  This is a case where a silent error could really waste a users time someday - and we should avoid this if possible.
> 
> In the current structure we will likely be able to have both performance and numerical correctness.  I think this is the best of all worlds

Yes but:

- We had only one contrastive network so by the priority/resource constrain point of view there is something wrong about what API to prioritize

- They never thought to fallback on vmap cause they known the design of the tf.image API. If you check again they have constrained the ops on CPU to try at least to use (large?) multicore machines for the preprocessing part of the computation:
https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/augmenters/simclr.py#L142:L143
